src/main/java/com/redislabs/Connector.java

```diff
@@ -245,55 +240,79 @@ public HashMap<String, Object> map(HashMap<String, Object> r) throws Exception {
     synchronized (this.connector) {
       try {
         Session session = connector.getSession();
         Transaction transaction = session.beginTransaction();
         boolean isMerge = true;
-        
-        for(Map<String, Object> r: record) {
-          lastStreamId = new String((byte[])r.get("id"));
-          Map<String, Object> map = (Map<String, Object>) r.get("value");
-          String sourceName = (String)map.remove(SOURCE_STR);
-          
-          String event = (String)map.remove(EVENT_STR);
-          if(event.charAt(0) != 'd') {
-            if(!isMerge) {
+
+        for (int idx = 0; idx < record.size(); ++idx) {
+          Map<String, Object> r = record.get(idx);
+          lastStreamId = new String((byte[]) r.get("id"));
+          Map<String, Object> value = (Map<String, Object>) r.get("value");
+          Map<String, Object> map = value.entrySet().stream()
+                  .collect(Collectors.toMap(Map.Entry::getKey, entry -> entry.getValue()));
+          String sourceName = (String) map.remove(SOURCE_STR);
+
+          String event = (String) map.remove(EVENT_STR);
+          if (event.charAt(0) != 'd') {
+            if (!isMerge) {
               transaction.commit();
               session.clear();
               transaction = session.beginTransaction();
               isMerge = true;
+              lastCommittedIdx = idx - 1;
             }
-            session.merge((String)map.remove(ENTETY_NAME_STR), map);
-          }else {
-            if(isMerge) {
+            session.merge((String) map.remove(ENTETY_NAME_STR), map);
+          } else {
+            if (isMerge) {
               transaction.commit();
               session.clear();
               transaction = session.beginTransaction();
               isMerge = false;
+              lastCommittedIdx = idx - 1;
             }
             Source source = Source.getSource(sourceName);
-            Object o = session.get((String)map.remove(ENTETY_NAME_STR), (Serializable)map.get(source.getIdProperty().getName()));
+            Object o = session.get((String) map.remove(ENTETY_NAME_STR), (Serializable) map.get(source.getIdProperty().getName()));
             // o can be null on hdel that removed the last field
-            if(o != null) {
+            if (o != null) {
               session.delete(o);
             }
           }
         }
-        
+
         transaction.commit();
         session.clear();
-      }catch (Exception e) {
-        msg = String.format("Failed commiting transaction error='%s'", e.toString());
+      }
+      catch (TransactionException|JDBCConnectionException|ServiceException ex) {
+        msg = String.format("Failed committing transaction error='%s'", ex);
         GearsBuilder.log(msg, LogLevel.WARNING);
+        lastStreamId = null;
+        cause = ex;
         connector.closeSession();
+
+      }
+      catch (Exception e) {
+        msg = String.format("Failed committing transaction error='%s'", e);
```

Comment:
So there is no really need to set the message here right? And then no need to set it to NULL after? Also I do not see where you check the `errorsToDLQ` value?

Reply:
Right, I moved it to separate exceptions and forgot to put it back in. Just did